# Running all Classification Models to get the best predictions
## Then, moving on to Neural Networks to get "Better Predictions"

In [1]:
from pandas import read_csv
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from complete_keras_model import xy_custom
from complete_keras_model import clean_df_y
from dummies_bins_test_train_cv import cross_val_process
from sklearn.metrics import *
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import RidgeClassifier
from sklearn.neural_network import MLPClassifier
# from clean_chess_game_log import main_cleanup
# _, _, _ = main_cleanup('../data/dest.pgn')

Using TensorFlow backend.


In [ ]:
from sklearn.metrics import C

In [6]:
df = read_csv('../data/use_for_predictions.csv')
df = df.loc[df['result'] != 0.5].copy()
df.reset_index(inplace=True)
df.drop(columns=['index'], inplace=True)

df, y = clean_df_y(df)

## Best Scores:

### Run 1: 100 test set:
- Gaussian Process Classifier
    - __ELO Difference + Start Time + Day of the Week__
    - [37, 11][13, 39]
    - __76%__
- K Neighbors Classifier
    - __ELO Difference + Game Time__
    - [37, 11][14, 38]
    - __75%__
- K Neighbors Classifier
    - __ELO Difference + Start Time__
    - [33, 15][10, 42]
    - __75%__

### Run 2: 75 test set:
- Gaussian Process Classifier
    - __ELO Difference, Game Time__
    - [30, 9][10, 26]
    - __74.7%__
- Gaussian Process Classifier
    - __ELO Difference, Start Time, Day of the Week__
    - [30, 9][11, 25]
    - __73.3%__
- Random Forest Classifier
    - __ELO Difference, Day of the Week__
    - [29, 10][10, 26]
    - __73.3%__
- K Neighbors Classifier
    - __ELO Difference, Start Time, Game Time, Day of the Week__
    - [28, 11][ 9, 27]
    - __73.3%__
- K Neighbors Classifier
    - __ELO Difference, Start Time, Day of the Week__
    - [27, 12][ 8, 28]
    - __73.3%__
- K Neighbors Classifier
    - __ELO Difference, Game Time, Day of the Week__
    - [29, 10][10, 26]
    - __73.3%__

### Run 3: 50 test set:
- Gaussian Process Classifier
    - __ELO Difference, Game Time__
    - [23, 4][ 6, 17]
    - __80%__
- Random Forest Classifier
    - __ELO Difference, Game Time__
    - [22,  5][ 5, 18]
    - __80%__
- Linear Discriminant Analysis
    - __ELO Difference, Start Time__
    - [21, 6][ 5, 18] 
    - __78.0%__
- Gaussian Process Classifier
    - __ELO Difference__
    - [22, 5][ 6, 17]
    - __78%__
- SGD Classifier
    - __ELO Difference, Game Time__
    - [22, 5][ 6, 17]
    - __78%__
- SGD Classifier
    - __ELO Difference__
    - [22, 5][ 6, 17]
    - __78%__

### Run 4: 150 test set:
- Random Forest Classifier
    - __ELO Difference, Day of the Week__
    - [55, 16][18, 61]
    - __77.3%__
- Gaussian Process Classifier
    - __ELO Difference, Start Time, Day of the Week__
    - [56, 15][21, 58]
    - __76%__
- K Neighbors Classifier
    - __ELO Difference, Game Time, Day of the Week__
    - [52, 19][17, 62]
    - __76%__
- K Neighbors Classifier
    - __ELO Difference, Day of the Week__
    - [54, 17][19, 60]
    - __76%__

### Run 5: 200 test set:
- Gaussian Process Classifier
    - __ELO Difference, Game Time__
    - [74, 23][25, 78]
    - __76%__
- Gaussian Process Classifier
    - __ELO Difference, Start Time, Day of the Week__
    - [77, 20][29, 74]
    - __75.5%__

In [7]:
# They are sorted decsending by the number of columns
all_cols = [
    ['diff_bin', 'color', 'time_bin', 'game_time', 'weekday'],
    ['diff_bin', 'time_bin', 'game_time', 'weekday'],
    ['diff_bin', 'color', 'time_bin', 'weekday'],
    ['diff_bin', 'time_bin', 'weekday'],
    ['diff_bin', 'color', 'time_bin', 'game_time'],
    ['diff_bin', 'time_bin', 'game_time'],
    ['diff_bin', 'color', 'time_bin'],
    ['diff_bin', 'time_bin'],
    ['diff_bin', 'color', 'game_time', 'weekday'],
    ['diff_bin', 'game_time', 'weekday'],
    ['diff_bin', 'color', 'weekday'],
    ['diff_bin', 'weekday'],
    ['diff_bin', 'color', 'game_time'],
    ['diff_bin', 'game_time'],
    ['diff_bin', 'color'],
    ['diff_bin']]

In [37]:
# Creating an empty dictionary to store all results
results = {}

### Linear Discriminant Analysis
### Run 1: 100 test set:
- 'diff_bin', 'color', 'time_bin', 'weekday' - [32, 16][13, 39] - 71%
- 'diff_bin', 'time_bin' - [31, 17][13, 39] - 70%

### Run 2: 75 test set:
- 'diff_bin', 'color', 'time_bin', 'game_time' - [25, 14][ 7, 29] - 72.0%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [25, 14][ 8, 28] - 70.7%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [26, 13][ 9, 27] - 70.7%
- 'diff_bin', 'time_bin', 'weekday' - [25, 14][ 8, 28] - 70.7%
- 'diff_bin', 'time_bin' - [25, 14][ 8, 28] - 70.7%

### Run 3: 50 test set:
- 'diff_bin', 'time_bin' - [21,  6][ 5, 18] - 78.0%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [20,  7][ 5, 18] - 76.0%
- 'diff_bin', 'game_time' - [21,  6][ 6, 17] - 76.0%
- 'diff_bin' - [21,  6][ 6, 17] - 76.0%
- 'diff_bin', 'time_bin', 'weekday' - [20,  7][ 5, 18] - 76.0%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [20,  7][ 6, 17] - 74.0%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [20,  7][ 6, 17] - 74.0%
- 'diff_bin', 'color', 'time_bin' - [20,  7][ 6, 17] - 74.0%
- 'diff_bin', 'color', 'game_time', 'weekday' - [20,  7][ 6, 17] - 74.0%
- 'diff_bin', 'color', 'weekday' - [20,  7][ 6, 17] - 74.0%
- 'diff_bin', 'weekday' - [20,  7][ 6, 17] - 74.0%
- 'diff_bin', 'color', 'game_time' - [20,  7][ 6, 17] - 74.0%
- 'diff_bin', 'color' - [20,  7][ 6, 17] - 74.0%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [18,  9][ 5, 18] - 72.0%
- 'diff_bin', 'time_bin', 'game_time' - [18,  9][ 5, 18] - 72.0%
- 'diff_bin', 'game_time', 'weekday' - [19,  8][ 6, 17] - 72.0%

### Run 4: 150 test set:
- diff_bin', 'color', 'time_bin', 'weekday' - [47, 24][16, 63] - 73.3%
- diff_bin', 'color', 'time_bin' - [46, 25][15, 64] - 73.3%
- diff_bin', 'time_bin', 'weekday' - [46, 25][17, 62] - 72.0%
- diff_bin', 'time_bin' - [47, 24][18, 61] - 72.0%
- diff_bin', 'color', 'time_bin', 'game_time' - [45, 26][18, 61] - 70.7%
- diff_bin', 'time_bin', 'game_time' - [44, 27][17, 62] - 70.7%
- diff_bin' - [45, 26][18, 61] - 70.7%
- diff_bin', 'time_bin', 'game_time', 'weekday' - [44, 27][18, 61] - 70.0%
- diff_bin', 'color', 'game_time' - [45, 26][19, 60] - 70.0%
- diff_bin', 'game_time' - [45, 26][19, 60] - 70.0%
- diff_bin', 'color' - [45, 26][19, 60] - 70.0%

### Run 5: 200 test set:

- 'diff_bin', 'color', 'time_bin', 'weekday' - [67, 30][24, 79] - 73.0%
- 'diff_bin', 'color', 'time_bin' - [66, 31][23, 80] - 73.0%
- 'diff_bin', 'time_bin', 'weekday' - [66, 31][25, 78] - 72.0%
- 'diff_bin', 'time_bin' - [65, 32][24, 79] - 72.0%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [66, 31][26, 77] - 71.5%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [65, 32][25, 78] - 71.5%
- 'diff_bin', 'game_time' - [64, 33][25, 78] - 71.0%
- 'diff_bin' - [64, 33][25, 78] - 71.0%
- 'diff_bin', 'color', 'game_time' - [64, 33][26, 77] - 70.5%
- 'diff_bin', 'color' - [64, 33][26, 77] - 70.5%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [62, 35][25, 78] - 70.0%
- 'diff_bin', 'weekday' - [65, 32][28, 75] - 70.0%

In [49]:
for clm in all_cols:
    X_train, X_test, y_train, y_test, X = xy_custom(df, y, 200, clm)
    LDA_clf = LinearDiscriminantAnalysis(solver='lsqr', shrinkage='auto'
                                        ).fit(X_train, y_train)
    cross_val_process(LDA_clf, X_train, y_train, X_test, y_test, cv=11)
    y_pred = LDA_clf.predict(X_test)
    y_pred = (y_pred > 0.5)
    cm = confusion_matrix(y_test, y_pred)
    results[(f'{clm}')] = [
        cm, (f'{((cm[0][0]+cm[1][1])/cm.sum()*100).round(1)}%')]
print(results)

y Shape: (2127,)
X Shape: (2127, 30)
X_train Shape: (1927, 30)
X_test Shape: (200, 30)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=76.4%
Scores(average_precision)=[0.742 0.77  0.783 0.719 0.745 0.756 0.731 0.828 0.797 0.743 0.789]
Standard_Deviation=0.031

Prediction_Confusion_Matrix:
[66|31]    	                                
[26|77]
Prediction_Accuracy=71.5%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.68      0.70        97
           1       0.71      0.75      0.73       103

   micro avg       0.71      0.71      0.71       200
   macro avg       0.72      0.71      0.71       200
weighted avg       0.72      0.71      0.71       200


y Shape: (2127,)
X Shape: (2127, 29)
X_train Shape: (1927, 29)
X_test Shape: (200, 29)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=76.2%
Scores(average_precision)=[0.741 0.778 0.784 0.724 

Average_Accuracy(average_precision)=76.78%
Scores(average_precision)=[0.743 0.793 0.787 0.717 0.743 0.745 0.736 0.833 0.784 0.779 0.786]
Standard_Deviation=0.032

Prediction_Confusion_Matrix:
[64|33]    	                                
[25|78]
Prediction_Accuracy=71.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.66      0.69        97
           1       0.70      0.76      0.73       103

   micro avg       0.71      0.71      0.71       200
   macro avg       0.71      0.71      0.71       200
weighted avg       0.71      0.71      0.71       200


y Shape: (2127,)
X Shape: (2127, 2)
X_train Shape: (1927, 2)
X_test Shape: (200, 2)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=76.77%
Scores(average_precision)=[0.751 0.79  0.778 0.704 0.755 0.745 0.746 0.836 0.795 0.764 0.783]
Standard_Deviation=0.033

Prediction_Confusion_Matrix:
[64|33]    	                      

### Gaussian Process Classifier
### Run 1: 100 test set:
- 'diff_bin', 'time_bin', 'weekday' - [37, 11][13, 39] - 76%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [35, 13][13, 39] - 74%
- 'diff_bin', 'color', 'time_bin' - [34, 14][14, 38] - 72%
- 'diff_bin', 'game_time' - [33, 15][13, 39] - 72%
- 'diff_bin', 'weekday' - [36, 12][17, 35] - 71%
- 'diff_bin', 'game_time', 'weekday' - [35, 13][16, 36] - 71%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [31, 17][13, 39] - 70%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [33, 15][15, 37] - 70%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [33, 15][15, 37] - 70%
- 'diff_bin', 'time_bin', 'game_time' - [31, 17][13, 39] - 70%
- 'diff_bin', 'time_bin' - [32, 16][14, 38] - 70%
- 'diff_bin', 'color', 'weekday' - [33, 15][15, 37] - 70%

### Run 2: 75 test set:
- 'diff_bin', 'game_time' - [30,  9][10, 26] - 74.7%
- 'diff_bin', 'time_bin', 'weekday' - [30,  9][11, 25] - 73.3%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [29, 10][11, 25] - 72.0%

### Run 3: 50 test set:
- 'diff_bin', 'game_time' - [23,  4][ 6, 17] - 80%
- 'diff_bin' - [22,  5][ 6, 17] - 78%
- 'diff_bin', 'time_bin', 'weekday' - [23,  4][ 8, 15] - 76%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [21,  6][ 8, 15] - 72%
- 'diff_bin', 'time_bin', 'game_time' - [20,  7][ 7, 16] - 72%
- 'diff_bin', 'time_bin' - [22,  5][ 9, 14] - 72%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [19,  8][ 7, 16] - 70%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [20,  7][ 8, 15] - 70%

### Run 4: 150 test set:
- 'diff_bin', 'time_bin', 'weekday' - [56, 15][21, 58] - 76%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [54, 17][20, 59] - 75.3%
- 'diff_bin', 'color', 'weekday' - [50, 21][16, 63] - 75.3%
- 'diff_bin', 'game_time', 'weekday' - [54, 17][21, 58] - 74.7%
- 'diff_bin', 'weekday' - [56, 15][23, 56] - 74.7%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [50, 21][19, 60] - 73.3%
- 'diff_bin', 'color', 'game_time', 'weekday' - [48, 23][18, 61] - 72.7%
- 'diff_bin', 'time_bin', 'game_time' - [50, 21][20, 59] - 72.7%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [51, 20][22, 57] - 72%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [50, 21][21, 58] - 72%
- 'diff_bin', 'time_bin' - [52, 19][23, 56] - 72%
- 'diff_bin' - [47, 24][18, 61] - 72%
- 'diff_bin', 'color', 'time_bin' - [52, 19][24, 55] - 71.3%
- 'diff_bin', 'game_time' - [49, 22][21, 58] - 71.3%

### Run 5: 200 test set:
- 'diff_bin', 'game_time' - [74, 23][25, 78] - 76%
- 'diff_bin', 'time_bin', 'weekday' - [77, 20][29, 74] - 75.5%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [75, 22][29, 74] - 74.5%
- 'diff_bin' - [68, 29][24, 79] - 73.5%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [72, 25][29, 74] - 73%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [73, 24][30, 73] - 73%
- 'diff_bin', 'game_time', 'weekday' - [74, 23][31, 72] - 73%
- 'diff_bin', 'weekday' - [75, 22][32, 71] - 73%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [73, 24][31, 72] - 72.5%
- 'diff_bin', 'color', 'weekday' - [67, 30][25, 78] - 72.5%
- 'diff_bin', 'color', 'game_time', 'weekday' - [68, 29][28, 75] - 71.5%
- 'diff_bin', 'color', 'time_bin' - [74, 23][35, 68] - 71%
- 'diff_bin', 'time_bin' - [71, 26][33, 70] - 70.5%
- 'diff_bin', 'color', 'game_time' - [70, 27][32, 71] - 70.5%
- 'diff_bin', 'time_bin', 'game_time' - [67, 30][30, 73] - 70%

In [68]:
for clm in all_cols:
    X_train, X_test, y_train, y_test, X = xy_custom(df, y, 200, clm)
    gpc_rbf_clf = GaussianProcessClassifier(n_jobs=-2, n_restarts_optimizer=10,
                                            random_state=9).fit(X_train, y_train)
    cross_val_process(gpc_rbf_clf, X_train, y_train, X_test, y_test, cv=11)
    y_pred = gpc_rbf_clf.predict(X_test)
    y_pred = (y_pred > 0.5)
    cm = confusion_matrix(y_test, y_pred)
    results[(f'{clm}')] = [
        cm, (f'{((cm[0][0]+cm[1][1])/cm.sum()*100).round(1)}%')]
print(results)

y Shape: (2127,)
X Shape: (2127, 30)
X_train Shape: (2077, 30)
X_test Shape: (50, 30)
y_train Shape: (2077,)
y_test Shape: (50,)
Average_Accuracy(average_precision)=76.14%
Scores(average_precision)=[0.742 0.74  0.801 0.714 0.738 0.741 0.784 0.782 0.708 0.84  0.787]
Standard_Deviation=0.038

Prediction_Confusion_Matrix:
[19|8]    	                                
[7|16]
Prediction_Accuracy=70.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.73      0.70      0.72        27
           1       0.67      0.70      0.68        23

   micro avg       0.70      0.70      0.70        50
   macro avg       0.70      0.70      0.70        50
weighted avg       0.70      0.70      0.70        50


y Shape: (2127,)
X Shape: (2127, 29)
X_train Shape: (2077, 29)
X_test Shape: (50, 29)
y_train Shape: (2077,)
y_test Shape: (50,)
Average_Accuracy(average_precision)=75.97%
Scores(average_precision)=[0.734 0.746 0.802 0.711 0.73

Average_Accuracy(average_precision)=76.44%
Scores(average_precision)=[0.753 0.786 0.752 0.717 0.712 0.734 0.797 0.775 0.731 0.83  0.821]
Standard_Deviation=0.039

Prediction_Confusion_Matrix:
[22|5]    	                                
[12|11]
Prediction_Accuracy=66.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.65      0.81      0.72        27
           1       0.69      0.48      0.56        23

   micro avg       0.66      0.66      0.66        50
   macro avg       0.67      0.65      0.64        50
weighted avg       0.67      0.66      0.65        50


y Shape: (2127,)
X Shape: (2127, 4)
X_train Shape: (2077, 4)
X_test Shape: (50, 4)
y_train Shape: (2077,)
y_test Shape: (50,)
Average_Accuracy(average_precision)=76.2%
Scores(average_precision)=[0.739 0.769 0.767 0.71  0.751 0.73  0.787 0.803 0.727 0.822 0.777]
Standard_Deviation=0.033

Prediction_Confusion_Matrix:
[19|8]    	                           

Average_Accuracy(average_precision)=75.89%
Scores(average_precision)=[0.752 0.775 0.773 0.669 0.753 0.735 0.789 0.753 0.713 0.827 0.808]
Standard_Deviation=0.042

Prediction_Confusion_Matrix:
[50|21]    	                                
[21|58]
Prediction_Accuracy=72.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.70      0.70      0.70        71
           1       0.73      0.73      0.73        79

   micro avg       0.72      0.72      0.72       150
   macro avg       0.72      0.72      0.72       150
weighted avg       0.72      0.72      0.72       150


y Shape: (2127,)
X Shape: (2127, 23)
X_train Shape: (1977, 23)
X_test Shape: (150, 23)
y_train Shape: (1977,)
y_test Shape: (150,)
Average_Accuracy(average_precision)=75.76%
Scores(average_precision)=[0.743 0.775 0.787 0.664 0.742 0.721 0.8   0.751 0.724 0.829 0.796]
Standard_Deviation=0.044

Prediction_Confusion_Matrix:
[50|21]    	                   

Average_Accuracy(average_precision)=75.57%
Scores(average_precision)=[0.756 0.79  0.75  0.679 0.749 0.702 0.791 0.75  0.704 0.815 0.827]
Standard_Deviation=0.045

Prediction_Confusion_Matrix:
[47|24]    	                                
[18|61]
Prediction_Accuracy=72.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.66      0.69        71
           1       0.72      0.77      0.74        79

   micro avg       0.72      0.72      0.72       150
   macro avg       0.72      0.72      0.72       150
weighted avg       0.72      0.72      0.72       150


{"['diff_bin', 'color', 'time_bin', 'game_time', 'weekday']": [array([[50, 21],
       [19, 60]]), '73.3%'], "['diff_bin', 'time_bin', 'game_time', 'weekday']": [array([[54, 17],
       [20, 59]]), '75.3%'], "['diff_bin', 'color', 'time_bin', 'weekday']": [array([[51, 20],
       [22, 57]]), '72.0%'], "['diff_bin', 'time_bin', 'weekday']": [array([[56,

Average_Accuracy(average_precision)=75.19%
Scores(average_precision)=[0.758 0.748 0.771 0.687 0.693 0.768 0.741 0.77  0.745 0.776 0.814]
Standard_Deviation=0.035

Prediction_Confusion_Matrix:
[68|29]    	                                
[28|75]
Prediction_Accuracy=71.5%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.71      0.70      0.70        97
           1       0.72      0.73      0.72       103

   micro avg       0.71      0.71      0.71       200
   macro avg       0.71      0.71      0.71       200
weighted avg       0.71      0.71      0.71       200


y Shape: (2127,)
X Shape: (2127, 9)
X_train Shape: (1927, 9)
X_test Shape: (200, 9)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=75.45%
Scores(average_precision)=[0.745 0.769 0.778 0.702 0.685 0.751 0.726 0.778 0.752 0.785 0.828]
Standard_Deviation=0.038

Prediction_Confusion_Matrix:
[74|23]    	                      

### Logistic Regression
### Run 1: 100 test set:
- 'diff_bin', 'color', 'weekday' - [32, 16][14, 38] - 70%
- 'diff_bin', 'color', 'game_time' - [32, 16][14, 38] - 70%
- 'diff_bin', 'color' - [32, 16][14, 38] - 70%

### Run 2: 75 test set:
- 'diff_bin', 'color', 'weekday' - [27, 12][ 9, 27] - 72.0%
- 'diff_bin', 'color', 'game_time' - [27, 12][ 9, 27] - 72.0%
- 'diff_bin', 'color' - [27, 12][ 9, 27] - 72.0%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [26, 13][ 9, 27] - 70.7%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [26, 13][ 9, 27] - 70.7%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [26, 13][ 9, 27] - 70.7%
- 'diff_bin', 'color', 'time_bin' - [26, 13][ 9, 27] - 70.7%
- 'diff_bin', 'color', 'game_time', 'weekday' - [26, 13][ 9, 27] - 70.7%

### Run 3: 50 test set:
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'time_bin', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'time_bin', 'game_time' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'color', 'time_bin' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'time_bin' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'color', 'game_time', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'game_time', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'color', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'color', 'game_time' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'game_time' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'color' - [21,  6][ 6, 17] - 76%
- 'diff_bin' - [21,  6][ 6, 17] - 76%

### Run 4: 150 test set:
- diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [46, 25][18, 61] - 71.3%
- diff_bin', 'color', 'time_bin', 'weekday' - [46, 25][18, 61] - 71.3%
- diff_bin', 'color', 'game_time', 'weekday' - [46, 25][18, 61] - 71.3%
- diff_bin', 'color', 'weekday' - [46, 25][18, 61] - 71.3%
- diff_bin', 'color' - [47, 24][19, 60] - 71.3%
- diff_bin', 'time_bin', 'game_time' - [45, 26][18, 61] - 70.7%
- diff_bin', 'color', 'time_bin' - [46, 25][19, 60] - 70.7%
- diff_bin', 'color', 'time_bin', 'game_time' - [45, 26][18, 61] - 70.7%
- diff_bin', 'time_bin' - [45, 26][18, 61] - 70.7%
- diff_bin', 'game_time', 'weekday' - [45, 26][18, 61] - 70.7%
- diff_bin', 'weekday' - [45, 26][18, 61] - 70.7%
- diff_bin', 'color', 'game_time' - [45, 26][18, 61] - 70.7%
- diff_bin', 'game_time' - [45, 26][18, 61] - 70.7%
- diff_bin' - [45, 26][18, 61] - 70.7%
- diff_bin', 'time_bin', 'game_time', 'weekday' - [45, 26][19, 60] - 70.0%
- diff_bin', 'time_bin', 'weekday' - [45, 26][19, 60] - 70.0%

### Run 5: 200 test set:
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [65, 32][25, 78] - 71.5%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [65, 32][25, 78] - 71.5%
- 'diff_bin', 'color', 'game_time', 'weekday' - [65, 32][25, 78] - 71.5%
- 'diff_bin', 'color', 'weekday' - [65, 32][25, 78] - 71.5%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [64, 33][25, 78] - 71%
- 'diff_bin', 'time_bin', 'game_time' - [64, 33][25, 78] - 71%
- 'diff_bin', 'color', 'time_bin' - [65, 32][26, 77] - 71%
- 'diff_bin', 'time_bin' - [64, 33][25, 78] - 71%
- 'diff_bin', 'game_time', 'weekday' - [64, 33][25, 78] - 71%
- 'diff_bin', 'weekday' - [64, 33][25, 78] - 71%
- 'diff_bin', 'color', 'game_time' - [64, 33][25, 78] - 71%
- 'diff_bin', 'game_time' - [64, 33][25, 78] - 71%
- 'diff_bin', 'color' - [64, 33][25, 78] - 71%
- 'diff_bin' - [64, 33][25, 78] - 71%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [64, 33][26, 77] - 70.5%
- 'diff_bin', 'time_bin', 'weekday' - [64, 33][26, 77] - 70.5%

In [54]:
for clm in all_cols:
    X_train, X_test, y_train, y_test, X = xy_custom(df, y, 200, clm)
    lgst_reg_clf = LogisticRegression(penalty='l2', class_weight='balanced', random_state=9,
                                      max_iter=5000, C=1e-3, solver='lbfgs', n_jobs=8,
                                      multi_class='auto').fit(X_train, y_train)
    cross_val_process(lgst_reg_clf, X_train, y_train, X_test, y_test, cv=11)
    y_pred = lgst_reg_clf.predict(X_test)
    y_pred = (y_pred > 0.5)
    cm = confusion_matrix(y_test, y_pred)
    results[(f'{clm}')] = [
        cm, (f'{((cm[0][0]+cm[1][1])/cm.sum()*100).round(1)}%')]
print(results)

y Shape: (2127,)
X Shape: (2127, 30)
X_train Shape: (1927, 30)
X_test Shape: (200, 30)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=77.18%
Scores(average_precision)=[0.75  0.798 0.786 0.72  0.747 0.744 0.743 0.84  0.793 0.779 0.79 ]
Standard_Deviation=0.033

Prediction_Confusion_Matrix:
[65|32]    	                                
[25|78]
Prediction_Accuracy=71.5%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.67      0.70        97
           1       0.71      0.76      0.73       103

   micro avg       0.71      0.71      0.71       200
   macro avg       0.72      0.71      0.71       200
weighted avg       0.72      0.71      0.71       200


y Shape: (2127,)
X Shape: (2127, 29)
X_train Shape: (1927, 29)
X_test Shape: (200, 29)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=77.1%
Scores(average_precision)=[0.747 0.797 0.788 0.72 

Average_Accuracy(average_precision)=76.87%
Scores(average_precision)=[0.746 0.796 0.785 0.72  0.741 0.738 0.735 0.833 0.79  0.78  0.79 ]
Standard_Deviation=0.033

Prediction_Confusion_Matrix:
[64|33]    	                                
[25|78]
Prediction_Accuracy=71.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.66      0.69        97
           1       0.70      0.76      0.73       103

   micro avg       0.71      0.71      0.71       200
   macro avg       0.71      0.71      0.71       200
weighted avg       0.71      0.71      0.71       200


y Shape: (2127,)
X Shape: (2127, 4)
X_train Shape: (1927, 4)
X_test Shape: (200, 4)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=77.05%
Scores(average_precision)=[0.746 0.797 0.785 0.72  0.745 0.745 0.741 0.838 0.79  0.779 0.788]
Standard_Deviation=0.032

Prediction_Confusion_Matrix:
[64|33]    	                      

### Ada Boost Classifier
### Run 1: 100 test set:
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [26, 22][ 7, 45] - 71%
- 'diff_bin', 'time_bin', 'weekday' - [26, 22][ 7, 45] - 71%
- 'diff_bin', 'time_bin', 'game_time' - [26, 22][ 7, 45] - 71%
- 'diff_bin', 'time_bin' - [26, 22][ 7, 45] - 71%
- 'diff_bin', 'game_time', 'weekday' - [26, 22][ 7, 45] - 71%
- 'diff_bin', 'weekday' - [26, 22][ 7, 45] - 71%
- 'diff_bin', 'game_time' - [26, 22][ 7, 45] - 71%
- 'diff_bin' - [26, 22][ 7, 45] - 71%


### Run 2: 75 test set:
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [23, 16][ 5, 31] - 72.0%
- 'diff_bin', 'time_bin', 'weekday' - [23, 16][ 5, 31] - 72.0%
- 'diff_bin', 'time_bin', 'game_time' - [23, 16][ 5, 31] - 72.0%
- 'diff_bin', 'time_bin' - [23, 16][ 5, 31] - 72.0%
- 'diff_bin', 'game_time', 'weekday' - [23, 16][ 5, 31] - 72.0%
- 'diff_bin', 'weekday' - [23, 16][ 5, 31] - 72.0%
- 'diff_bin', 'game_time' - [23, 16][ 5, 31] - 72.0%
- 'diff_bin' - [23, 16][ 5, 31] - 72.0%

### Run 3: 50 test set:
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [19,  8][ 4, 19] - 76%
- 'diff_bin', 'time_bin', 'weekday' - [19,  8][ 4, 19] - 76%
- 'diff_bin', 'time_bin', 'game_time' - [19,  8][ 4, 19] - 76%
- 'diff_bin', 'time_bin' - [19,  8][ 4, 19] - 76%
- 'diff_bin', 'game_time', 'weekday' - [19,  8][ 4, 19] - 76%
- 'diff_bin', 'weekday' - [19,  8][ 4, 19] - 76%
- 'diff_bin', 'game_time' - [19,  8][ 4, 19] - 76%
- 'diff_bin' - [19,  8][ 4, 19] - 76%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [20,  7][ 7, 16] - 72%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [20,  7][ 7, 16] - 72%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [20,  7][ 7, 16] - 72%
- 'diff_bin', 'color', 'time_bin' - [20,  7][ 7, 16] - 72%
- 'diff_bin', 'color', 'game_time', 'weekday' - [20,  7][ 7, 16] - 72%
- 'diff_bin', 'color', 'weekday' - [20,  7][ 7, 16] - 72%
- 'diff_bin', 'color', 'game_time' - [20,  7][ 7, 16] - 72%
- 'diff_bin', 'color' - [20,  7][ 7, 16] - 72%

### Run 4: 150 test set:
- 'diff_bin', 'game_time', 'weekday' - [41, 30][ 7, 72] - 75.3%
- 'diff_bin', 'weekday' - [41, 30][ 7, 72] - 75.3%
- 'diff_bin', 'game_time' - [41, 30][ 7, 72] - 75.3%
- 'diff_bin' - [41, 30][ 7, 72] - 75.3%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [46, 25][16, 63] - 72.7%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [46, 25][16, 63] - 72.7%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [46, 25][16, 63] - 72.7%
- 'diff_bin', 'color', 'time_bin' - [46, 25][16, 63] - 72.7%
- 'diff_bin', 'color', 'game_time', 'weekday' - [46, 25][16, 63] - 72.7%
- 'diff_bin', 'color', 'weekday' - [46, 25][16, 63] - 72.7%
- 'diff_bin', 'color', 'game_time' - [46, 25][16, 63] - 72.7%
- 'diff_bin', 'color' - [46, 25][16, 63] - 72.7%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [51, 20][23, 56] - 71.3%
- 'diff_bin', 'time_bin', 'weekday' - [51, 20][23, 56] - 71.3%
- 'diff_bin', 'time_bin', 'game_time' - [51, 20][23, 56] - 71.3%
- 'diff_bin', 'time_bin' - [51, 20][23, 56] - 71.3%

### Run 5: 200 test set:
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [64, 33][20, 83] - 73.5%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [64, 33][20, 83] - 73.5%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [64, 33][20, 83] - 73.5%
- 'diff_bin', 'color', 'time_bin' - [64, 33][20, 83] - 73.5%
- 'diff_bin', 'color', 'game_time', 'weekday' - [65, 32][22, 81] - 73.0%
- 'diff_bin', 'game_time', 'weekday' - [55, 42][12, 91] - 73.0%
- 'diff_bin', 'color', 'weekday' - [65, 32][22, 81] - 73.0%
- 'diff_bin', 'weekday' - [55, 42][12, 91] - 73.0%
- 'diff_bin', 'color', 'game_time' - [65, 32][22, 81] - 73.0%
- 'diff_bin', 'game_time' - [55, 42][12, 91] - 73.0%
- 'diff_bin', 'color' - [65, 32][22, 81] - 73.0%
- 'diff_bin' - [55, 42][12, 91] - 73.0%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [57, 40][15, 88] - 72.5%
- 'diff_bin', 'time_bin', 'weekday' - [57, 40][15, 88] - 72.5%
- 'diff_bin', 'time_bin', 'game_time' - [57, 40][15, 88] - 72.5%
- 'diff_bin', 'time_bin' - [57, 40][15, 88] - 72.5%

In [73]:
for clm in all_cols:
    X_train, X_test, y_train, y_test, X = xy_custom(df, y, 200, clm)
    ada_clf = AdaBoostClassifier(n_estimators=274, learning_rate=0.013,
                                 random_state=9).fit(X_train, y_train)
    cross_val_process(ada_clf, X_train, y_train, X_test, y_test, cv=11)
    y_pred = ada_clf.predict(X_test)
    y_pred = (y_pred > 0.5)
    cm = confusion_matrix(y_test, y_pred)
    results[(f'{clm}')] = [
        cm, (f'{((cm[0][0]+cm[1][1])/cm.sum()*100).round(1)}%')]
print(results)

y Shape: (2127,)
X Shape: (2127, 30)
X_train Shape: (1927, 30)
X_test Shape: (200, 30)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=75.32%
Scores(average_precision)=[0.764 0.752 0.763 0.687 0.734 0.744 0.743 0.813 0.746 0.759 0.781]
Standard_Deviation=0.029

Prediction_Confusion_Matrix:
[64|33]    	                                
[20|83]
Prediction_Accuracy=73.5%
Feature importance = [0.91605839 0.0620438  0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.01459854 0.         0.         0.         0.00729927
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.         0.        ]

Classification Report:
              precision    recall  f1-score   support

           0       0.76      0.66      0.71        97
           1       0.72      0.81      0.76       103

   micro avg       0.73      0.73      0.73       200
   macro avg       0.74

Average_Accuracy(average_precision)=75.22%
Scores(average_precision)=[0.759 0.755 0.761 0.687 0.73  0.744 0.74  0.809 0.749 0.757 0.781]
Standard_Deviation=0.029

Prediction_Confusion_Matrix:
[65|32]    	                                
[22|81]
Prediction_Accuracy=73.0%
Feature importance = [0.9270073 0.0729927 0.        0.        0.        0.        0.
 0.        0.        0.       ]

Classification Report:
              precision    recall  f1-score   support

           0       0.75      0.67      0.71        97
           1       0.72      0.79      0.75       103

   micro avg       0.73      0.73      0.73       200
   macro avg       0.73      0.73      0.73       200
weighted avg       0.73      0.73      0.73       200


y Shape: (2127,)
X Shape: (2127, 9)
X_train Shape: (1927, 9)
X_test Shape: (200, 9)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=74.33%
Scores(average_precision)=[0.729 0.779 0.758 0.68  0.705 0.723 0.705 0.804 0.746 0.764 0.

### SGD Classifier
### Run 1: 100 test set:
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [28, 20][ 8, 44] - 72%
- 'diff_bin', 'color', 'time_bin' - [29, 19][10, 42] - 71%
- 'diff_bin', 'color', 'weekday' - [29, 19][10, 42] - 71%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [30, 18][12, 40] - 70%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [29, 19][11, 41] - 70%
- 'diff_bin', 'time_bin' - [31, 17][13, 39] - 70%
- 'diff_bin', 'game_time', 'weekday' - [26, 22][ 8, 44] - 70%


### Run 2: 75 test set:
- 'diff_bin', 'weekday' - [23, 16[ 5, 31] - 72.0%
- 'diff_bin', 'color', 'game_time' - [26, 13[ 8, 28] - 72.0%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [23, 16[ 6, 30] - 70.7%
- 'diff_bin', 'time_bin', 'weekday' - [24, 15[ 7, 29] - 70.7%
- 'diff_bin', 'color', 'time_bin' - [25, 14[ 8, 28] - 70.7%
- 'diff_bin', 'time_bin' - [24, 15[ 7, 29] - 70.7%
- 'diff_bin', 'game_time', 'weekday' - [22, 17[ 5, 31] - 70.7%
- 'diff_bin', 'game_time' - [23, 16[ 6, 30] - 70.7%

### Run 3: 50 test set:
- 'diff_bin', 'game_time' - [22,  5][ 6, 17] - 78%
- 'diff_bin' - [22,  5][ 6, 17] - 78%
- 'diff_bin', 'time_bin' - [20,  7][ 5, 18] - 76%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [22,  5][ 8, 15] - 74%
- 'diff_bin', 'time_bin', 'game_time' - [21,  6][ 7, 16] - 74%
- 'diff_bin', 'color', 'time_bin' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'game_time', 'weekday' - [21,  6][ 7, 16] - 74%
- 'diff_bin', 'weekday' - [22,  5][ 8, 15] - 74%
- 'diff_bin', 'color', 'game_time' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'color' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'time_bin', 'weekday' - [21,  6][ 8, 15] - 72%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [19,  8][ 6, 17] - 72%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [20,  7][ 8, 15] - 70%
- 'diff_bin', 'color', 'weekday' - [20,  7][ 8, 15] - 70%

### Run 4: 150 test set:
- 'diff_bin' - [50, 21][20, 59] - 72.7%
- 'diff_bin', 'game_time', 'weekday' - [54, 17][25, 54] - 72%
- 'diff_bin', 'weekday' - [52, 19][23, 56] - 72%
- 'diff_bin', 'color', 'time_bin' - [47, 24][19, 60] - 71.3%
- 'diff_bin', 'time_bin' - [48, 23][20, 59] - 71.3%
- 'diff_bin', 'game_time' - [50, 21][23, 56] - 70.7%
- 'diff_bin', 'color' - [50, 21][23, 56] - 70.7%
- 'diff_bin', 'time_bin', 'weekday' - [48, 23][21, 58] - 70.7%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [50, 21][24, 55] - 70%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [45, 26][19, 60] - 70%
- 'diff_bin', 'time_bin', 'game_time' - [44, 27][18, 61] - 70%
- 'diff_bin', 'color', 'game_time' - [45, 26][19, 60] - 70%

### Run 5: 200 test set:
- 'diff_bin', 'game_time', 'weekday' - [80, 17][42, 61] - 70.5%

In [58]:
for clm in all_cols:
    X_train, X_test, y_train, y_test, X = xy_custom(df, y, 200, clm)
    SGD_clf = SGDClassifier(loss='hinge', penalty='l2', max_iter=1e3,
                            shuffle=False, n_jobs=8, random_state=9,
                            class_weight='balanced').fit(X_train, y_train)
    cross_val_process(SGD_clf, X_train, y_train, X_test, y_test, cv=11)
    y_pred = SGD_clf.predict(X_test)
    y_pred = (y_pred > 0.5)
    cm = confusion_matrix(y_test, y_pred)
    results[(f'{clm}')] = [
        cm, (f'{((cm[0][0]+cm[1][1])/cm.sum()*100).round(1)}%')]
print(results)

y Shape: (2127,)
X Shape: (2127, 30)
X_train Shape: (1927, 30)
X_test Shape: (200, 30)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=76.38%
Scores(average_precision)=[0.758 0.777 0.768 0.702 0.751 0.754 0.735 0.811 0.797 0.757 0.791]
Standard_Deviation=0.029

Prediction_Confusion_Matrix:
[79|18]    	                                
[48|55]
Prediction_Accuracy=67.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.62      0.81      0.71        97
           1       0.75      0.53      0.62       103

   micro avg       0.67      0.67      0.67       200
   macro avg       0.69      0.67      0.67       200
weighted avg       0.69      0.67      0.66       200


y Shape: (2127,)
X Shape: (2127, 29)
X_train Shape: (1927, 29)
X_test Shape: (200, 29)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=75.91%
Scores(average_precision)=[0.755 0.785 0.761 0.70

Average_Accuracy(average_precision)=76.08%
Scores(average_precision)=[0.733 0.793 0.768 0.7   0.743 0.742 0.715 0.829 0.775 0.78  0.789]
Standard_Deviation=0.036

Prediction_Confusion_Matrix:
[81|16]    	                                
[49|54]
Prediction_Accuracy=67.5%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.62      0.84      0.71        97
           1       0.77      0.52      0.62       103

   micro avg       0.68      0.68      0.68       200
   macro avg       0.70      0.68      0.67       200
weighted avg       0.70      0.68      0.67       200


y Shape: (2127,)
X Shape: (2127, 4)
X_train Shape: (1927, 4)
X_test Shape: (200, 4)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=76.7%
Scores(average_precision)=[0.754 0.781 0.781 0.705 0.743 0.751 0.776 0.805 0.796 0.77  0.774]
Standard_Deviation=0.026

Prediction_Confusion_Matrix:
[79|18]    	                       

### Random Forest Classifier
### Run 1: 100 test set:
- 'diff_bin', 'color', 'time_bin', 'weekday' - [32, 16][11, 41] - 73%
- 'diff_bin', 'game_time' - [32, 16][11, 41] - 73%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [33, 15][13, 39] - 72%
- 'diff_bin', 'time_bin', 'game_time' - [34, 14][14, 38] - 72%
- 'diff_bin', 'game_time', 'weekday' - [32, 16][12, 40] - 72%
- 'diff_bin', 'weekday' - [32, 16][12, 40] - 72%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [31, 17][12, 40] - 71%
- 'diff_bin', 'time_bin', 'weekday' - [34, 14][15, 37] - 71%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [32, 16][13, 39] - 71%

### Run 2: 75 test set:
- 'diff_bin', 'weekday' - [29, 10][10, 26] - 73.3%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [26, 13][ 8, 28] - 72.0%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [27, 12][ 9, 27] - 72.0%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [27, 12][10, 26] - 70.7%
- 'diff_bin', 'time_bin', 'weekday' - [28, 11][11, 25] - 70.7%

### Run 3: 50 test set:
- 'diff_bin', 'game_time' - [22,  5][ 5, 18] - 80%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [21,  6][ 7, 16] - 74%
- 'diff_bin', 'time_bin', 'weekday' - [22,  5][ 8, 15] - 74%
- 'diff_bin', 'time_bin', 'game_time' - [21,  6][ 7, 16] - 74%
- 'diff_bin', 'weekday' - [22,  5][ 8, 15] - 74%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [19,  8][ 6, 17] - 72%
- 'diff_bin', 'game_time', 'weekday' - [21,  6][ 8, 15] - 72%
- 'diff_bin' - [22,  5][ 9, 14] - 72%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [19,  8][ 7, 16] - 70%

### Run 4: 150 test set:
- 'diff_bin', 'weekday' - [55, 16][18, 61] - 77.3%
- 'diff_bin', 'time_bin', 'game_time' - [51, 20][18, 61] - 74.7%
- 'diff_bin', 'game_time', 'weekday' - [51, 20][18, 61] - 74.7%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [52, 19][20, 59] - 74.0%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [48, 23][17, 62] - 73.3%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [52, 19][22, 57] - 72.7%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [49, 22][20, 59] - 72.0%
- 'diff_bin', 'game_time' - [46, 25][17, 62] - 72.0%
- 'diff_bin' - [47, 24][18, 61] - 72.0%
- 'diff_bin', 'time_bin', 'weekday' - [52, 19][24, 55] - 71.3%
- 'diff_bin', 'time_bin' - [47, 24][20, 59] - 70.7%
- 'diff_bin', 'color', 'time_bin' - [48, 23][22, 57] - 70.0%

### Run 5: 200 test set:
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [71, 26][26, 77] - 74.0%
- 'diff_bin', 'game_time', 'weekday' - [70, 27][25, 78] - 74.0%
- 'diff_bin', 'weekday' - [70, 27][26, 77] - 73.5%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [70, 27][28, 75] - 72.5%
- 'diff_bin', 'time_bin', 'game_time' - [70, 27][28, 75] - 72.5%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [69, 28][28, 75] - 72.0%
- 'diff_bin', 'game_time' - [58, 39][17, 86] - 72.0%
- 'diff_bin', 'time_bin', 'weekday' - [70, 27][30, 73] - 71.5%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [69, 28][30, 73] - 71.0%
- 'diff_bin' - [65, 32][26, 77] - 71.0%

In [77]:
for clm in all_cols:
    X_train, X_test, y_train, y_test, X = xy_custom(df, y, 200, clm)
    rand_frst_clf = RandomForestClassifier(n_estimators=1000, criterion='entropy', n_jobs=-2,
                                           min_samples_leaf=2, random_state=9,
                                           class_weight='balanced').fit(X_train, y_train)
    cross_val_process(rand_frst_clf, X_train, y_train, X_test, y_test, cv=11)
    y_pred = rand_frst_clf.predict(X_test)
    y_pred = (y_pred > 0.5)
    cm = confusion_matrix(y_test, y_pred)
    results[(f'{clm}')] = [
        cm, (f'{((cm[0][0]+cm[1][1])/cm.sum()*100).round(1)}%')]
print(results)

y Shape: (2127,)
X Shape: (2127, 30)
X_train Shape: (1927, 30)
X_test Shape: (200, 30)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=75.5%
Scores(average_precision)=[0.734 0.757 0.758 0.71  0.721 0.783 0.724 0.795 0.768 0.75  0.806]
Standard_Deviation=0.03

Prediction_Confusion_Matrix:
[69|28]    	                                
[28|75]
Prediction_Accuracy=72.0%
Feature importance = [5.90784183e-01 3.61066900e-02 3.14074164e-04 1.64543187e-03
 1.75550406e-03 4.43775675e-03 1.01924021e-02 1.14602855e-02
 1.46707713e-02 1.10925158e-02 1.19895293e-02 1.11597015e-02
 1.33434350e-02 1.37427033e-02 1.60333252e-02 1.32306552e-02
 1.32552613e-02 1.31799485e-02 1.12616910e-02 9.80641444e-03
 1.10805442e-02 1.14688186e-02 1.53532567e-02 1.98444652e-02
 2.49839842e-02 2.34867049e-02 2.13409828e-02 2.15965084e-02
 1.97319430e-02 2.16505128e-02]

Classification Report:
              precision    recall  f1-score   support

           0       0.71      0.71      0.

Average_Accuracy(average_precision)=72.7%
Scores(average_precision)=[0.719 0.756 0.723 0.632 0.715 0.733 0.717 0.747 0.733 0.744 0.779]
Standard_Deviation=0.035

Prediction_Confusion_Matrix:
[65|32]    	                                
[31|72]
Prediction_Accuracy=68.5%
Feature importance = [8.87009555e-01 5.84822534e-04 3.24029385e-03 2.45526862e-03
 3.45509581e-03 4.86800674e-03 6.25548496e-03 7.42165246e-03
 4.88648978e-03 5.95192925e-03 5.13185205e-03 5.37963997e-03
 9.13636347e-03 6.42488199e-03 5.80629360e-03 4.72459243e-03
 8.10234215e-03 6.80740235e-03 6.05141820e-03 7.64899404e-03
 8.65762039e-03]

Classification Report:
              precision    recall  f1-score   support

           0       0.68      0.67      0.67        97
           1       0.69      0.70      0.70       103

   micro avg       0.69      0.69      0.69       200
   macro avg       0.68      0.68      0.68       200
weighted avg       0.68      0.69      0.68       200


y Shape: (2127,)
X Shape: (2127, 10

### Ridge Classifier
### Run 1: 100 test set:
- 'diff_bin', 'color', 'time_bin' - [29, 19][11, 41] - 70%

### Run 2: 75 test set:
- 'diff_bin', 'color', 'time_bin', 'game_time' - [25, 14][ 8, 28] - 70.7%
- 'diff_bin', 'color', 'time_bin' - [25, 14][ 8, 28] - 70.7%

### Run 3: 50 test set:
- 'diff_bin', 'game_time' - [21,  6][ 6, 17] - 76%
- 'diff_bin' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'color', 'time_bin' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'color', 'game_time', 'weekday' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'color', 'weekday' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'color', 'game_time' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'color' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [18,  9][ 5, 18] - 72%
- 'diff_bin', 'time_bin', 'weekday' - [18,  9][ 5, 18] - 72%
- 'diff_bin', 'time_bin', 'game_time' - [18,  9][ 5, 18] - 72%
- 'diff_bin', 'time_bin' - [18,  9][ 5, 18] - 72%
- 'diff_bin', 'game_time', 'weekday' - [19,  8][ 6, 17] - 72%
- 'diff_bin', 'weekday' - [19,  8][ 6, 17] - 72%

### Run 4: 150 test set:
- 'diff_bin', 'color', 'time_bin' - [45, 26][15, 64] - 72.7%
- 'diff_bin', 'color' - [45, 26][17, 62] - 71.3%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [45, 26][17, 62] - 71.3%
- 'diff_bin', 'time_bin', 'weekday' - [44, 27][17, 62] - 70.7%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [45, 26][18, 61] - 70.7%
- 'diff_bin', 'time_bin' - [45, 26][18, 61] - 70.7%
- 'diff_bin', 'game_time' - [45, 26][18, 61] - 70.7%
- 'diff_bin' - [45, 26][18, 61] - 70.7%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [45, 26][19, 60] - 70.0%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [43, 28][17, 62] - 70.0%
- 'diff_bin', 'time_bin', 'game_time' - [43, 28][17, 62] - 70.0%
- 'diff_bin', 'color', 'weekday' - [45, 26][19, 60] - 70.0%
- 'diff_bin', 'color', 'game_time' - [45, 26][19, 60] - 70.0%

### Run 5: 200 test set:
- 'diff_bin', 'color' - [64, 33][23, 80] - 72.0%
- 'diff_bin', 'color', 'time_bin' - [64, 33][24, 79] - 71.5%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [65, 32][26, 77] - 71.0%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [64, 33][25, 78] - 71.0%
- 'diff_bin', 'time_bin', 'game_time' - [63, 34][24, 79] - 71.0%
- 'diff_bin', 'game_time' - [64, 33][25, 78] - 71.0%
- 'diff_bin' - [64, 33][25, 78] - 71.0%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [62, 35][24, 79] - 70.5%
- 'diff_bin', 'time_bin', 'weekday' - [63, 34][25, 78] - 70.5%
- 'diff_bin', 'time_bin' - [64, 33][26, 77] - 70.5%
- 'diff_bin', 'color', 'weekday' - [64, 33][26, 77] - 70.5%
- 'diff_bin', 'color', 'game_time' - [64, 33][26, 77] - 70.5%

In [62]:
for clm in all_cols:
    X_train, X_test, y_train, y_test, X = xy_custom(df, y, 200, clm)
    ridge_clf = RidgeClassifier(class_weight='balanced', random_state=9
                               ).fit(X_train, y_train)
    cross_val_process(ridge_clf, X_train, y_train, X_test, y_test, cv=11)
    y_pred = ridge_clf.predict(X_test)
    y_pred = (y_pred > 0.5)
    cm = confusion_matrix(y_test, y_pred)
    results[(f'{clm}')] = [
        cm, (f'{((cm[0][0]+cm[1][1])/cm.sum()*100).round(1)}%')]
print(results)

y Shape: (2127,)
X Shape: (2127, 30)
X_train Shape: (1927, 30)
X_test Shape: (200, 30)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=76.0%
Scores(average_precision)=[0.748 0.773 0.773 0.709 0.756 0.745 0.719 0.824 0.797 0.728 0.789]
Standard_Deviation=0.034

Prediction_Confusion_Matrix:
[65|32]    	                                
[30|73]
Prediction_Accuracy=69.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.68      0.67      0.68        97
           1       0.70      0.71      0.70       103

   micro avg       0.69      0.69      0.69       200
   macro avg       0.69      0.69      0.69       200
weighted avg       0.69      0.69      0.69       200


y Shape: (2127,)
X Shape: (2127, 29)
X_train Shape: (1927, 29)
X_test Shape: (200, 29)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=75.7%
Scores(average_precision)=[0.747 0.78  0.772 0.711 

Average_Accuracy(average_precision)=76.81%
Scores(average_precision)=[0.753 0.785 0.785 0.704 0.753 0.752 0.755 0.837 0.794 0.755 0.776]
Standard_Deviation=0.032

Prediction_Confusion_Matrix:
[64|33]    	                                
[26|77]
Prediction_Accuracy=70.5%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.71      0.66      0.68        97
           1       0.70      0.75      0.72       103

   micro avg       0.70      0.70      0.70       200
   macro avg       0.71      0.70      0.70       200
weighted avg       0.71      0.70      0.70       200


y Shape: (2127,)
X Shape: (2127, 3)
X_train Shape: (1927, 3)
X_test Shape: (200, 3)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=76.74%
Scores(average_precision)=[0.743 0.793 0.787 0.717 0.743 0.745 0.734 0.833 0.784 0.776 0.786]
Standard_Deviation=0.032

Prediction_Confusion_Matrix:
[64|33]    	                      

### K Neighbors Classifier
### Run 1: 100 test set:
- 'diff_bin', 'game_time' - [37, 11][14, 38] - 75%
- 'diff_bin', 'time_bin' - [33, 15][10, 42] - 75%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [35, 13][13, 39] - 74%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [34, 14][12, 40] - 74%
- 'diff_bin', 'time_bin', 'weekday' - [33, 15][11, 41] - 74%
- 'diff_bin', 'weekday' - [35, 13][13, 39] - 74%
- 'diff_bin' - [34, 14][13, 39] - 73%
- 'diff_bin', 'time_bin', 'game_time' - [33, 15][13, 39] - 72%
- 'diff_bin', 'game_time', 'weekday' - [35, 13][15, 37] - 72%
- 'diff_bin', 'color', 'weekday' - [36, 12][16, 36] - 72%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [32, 16][13, 39] - 71%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [32, 16][13, 39] - 71%
- 'diff_bin', 'color', 'game_time', 'weekday' - [33, 15][15, 37] - 70%


### Run 2: 75 test set:
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [28, 11][ 9, 27] - 73.3%
- 'diff_bin', 'time_bin', 'weekday' - [27, 12][ 8, 28] - 73.3%
- 'diff_bin', 'game_time', 'weekday' - [29, 10][10, 26] - 73.3%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [27, 12][ 9, 27] - 72.0%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [27, 12][10, 26] - 70.7%
- 'diff_bin', 'time_bin', 'game_time' - [26, 13][ 9, 27] - 70.7%
- 'diff_bin', 'color', 'time_bin' - [28, 11][11, 25] - 70.7%


### Run 3: 50 test set:
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [21,  6][ 6, 17] - 76%
- 'diff_bin', 'time_bin', 'weekday' - [22,  5][ 7, 16] - 76%
- 'diff_bin', 'game_time' - [22,  5][ 7, 16] - 76%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [21,  6][ 7, 16] - 74%
- 'diff_bin', 'time_bin' - [20,  7][ 6, 17] - 74%
- 'diff_bin', 'game_time', 'weekday' - [22,  5][ 9, 14] - 72%
- 'diff_bin', 'weekday' - [23,  4][10, 13] - 72%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [20,  7][ 8, 15] - 70%
- 'diff_bin', 'time_bin', 'game_time' - [18,  9][ 6, 17] - 70%
- 'diff_bin' - [21,  6][ 9, 14] - 70%

### Run 4: 150 test set:
- 'diff_bin', 'game_time', 'weekday' - [52, 19][17, 62] - 76.0%
- 'diff_bin', 'weekday' - [54, 17][19, 60] - 76.0%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [51, 20][17, 62] - 75.3%
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [55, 16][21, 58] - 75.3%
- 'diff_bin', 'time_bin' - [50, 21][18, 61] - 74.0%
- 'diff_bin', 'time_bin', 'weekday' - [52, 19][21, 58] - 73.3%
- 'diff_bin', 'color', 'game_time', 'weekday' - [49, 22][18, 61] - 73.3%
- 'diff_bin', 'time_bin', 'game_time' - [50, 21][20, 59] - 72.7%
- 'diff_bin', 'game_time' - [49, 22][19, 60] - 72.7%
- 'diff_bin', 'color', 'weekday' - [51, 20][22, 57] - 72.0%
- 'diff_bin' - [48, 23][19, 60] - 72.0%
- 'diff_bin', 'color', 'time_bin' - [48, 23][20, 59] - 71.3%
- 'diff_bin', 'color' - [45, 26][17, 62] - 71.3%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [51, 20][24, 55] - 70.7%
- 'diff_bin', 'color', 'game_time' - [43, 28][17, 62] - 70.0%

### Run 5: 200 test set:
- 'diff_bin', 'time_bin', 'game_time', 'weekday' - [77, 20][30, 73] - 75.0%
- 'diff_bin', 'color', 'time_bin', 'weekday' - [72, 25][25, 78] - 75.0%
- 'diff_bin', 'time_bin', 'weekday' - [73, 24][26, 77] - 75.0%
- 'diff_bin', 'weekday' - [73, 24][26, 77] - 75.0%
- 'diff_bin', 'time_bin' - [72, 25][26, 77] - 74.5%
- 'diff_bin', 'color', 'time_bin', 'game_time', 'weekday' - [71, 26][26, 77] - 74.0%
- 'diff_bin', 'game_time', 'weekday' - [72, 25][27, 76] - 74.0%
- 'diff_bin', 'time_bin', 'game_time' - [72, 25][28, 75] - 73.5%
- 'diff_bin', 'color', 'game_time', 'weekday' - [72, 25][30, 73] - 72.5%
- 'diff_bin', 'color', 'weekday' - [71, 26][29, 74] - 72.5%
- 'diff_bin', 'color', 'time_bin', 'game_time' - [73, 24][33, 70] - 71.5%
- 'diff_bin' - [67, 30][28, 75] - 71.0%
- 'diff_bin', 'color', 'time_bin' - [70, 27][33, 70] - 70.0%
- 'diff_bin', 'game_time' - [61, 36][24, 79] - 70.0%

In [66]:
for clm in all_cols:
    X_train, X_test, y_train, y_test, X = xy_custom(df, y, 200, clm)
    KNN_clf = KNeighborsClassifier(n_neighbors=19, leaf_size=88, n_jobs=8,
                                   metric='manhattan').fit(X_train, y_train)
    cross_val_process(KNN_clf, X_train, y_train, X_test, y_test, cv=11)
    y_pred = KNN_clf.predict(X_test)
    y_pred = (y_pred > 0.5)
    cm = confusion_matrix(y_test, y_pred)
    results[(f'{clm}')] = [
        cm, (f'{((cm[0][0]+cm[1][1])/cm.sum()*100).round(1)}%')]
print(results)

y Shape: (2127,)
X Shape: (2127, 30)
X_train Shape: (1927, 30)
X_test Shape: (200, 30)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=74.51%
Scores(average_precision)=[0.72  0.756 0.736 0.699 0.702 0.761 0.697 0.788 0.775 0.778 0.785]
Standard_Deviation=0.034

Prediction_Confusion_Matrix:
[71|26]    	                                
[26|77]
Prediction_Accuracy=74.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.73      0.73      0.73        97
           1       0.75      0.75      0.75       103

   micro avg       0.74      0.74      0.74       200
   macro avg       0.74      0.74      0.74       200
weighted avg       0.74      0.74      0.74       200


y Shape: (2127,)
X Shape: (2127, 29)
X_train Shape: (1927, 29)
X_test Shape: (200, 29)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=74.53%
Scores(average_precision)=[0.715 0.78  0.743 0.70

Average_Accuracy(average_precision)=74.88%
Scores(average_precision)=[0.732 0.797 0.755 0.711 0.711 0.73  0.688 0.766 0.761 0.81  0.776]
Standard_Deviation=0.036

Prediction_Confusion_Matrix:
[73|24]    	                                
[26|77]
Prediction_Accuracy=75.0%
No Feature Importances

Classification Report:
              precision    recall  f1-score   support

           0       0.74      0.75      0.74        97
           1       0.76      0.75      0.75       103

   micro avg       0.75      0.75      0.75       200
   macro avg       0.75      0.75      0.75       200
weighted avg       0.75      0.75      0.75       200


y Shape: (2127,)
X Shape: (2127, 4)
X_train Shape: (1927, 4)
X_test Shape: (200, 4)
y_train Shape: (1927,)
y_test Shape: (200,)
Average_Accuracy(average_precision)=73.77%
Scores(average_precision)=[0.725 0.759 0.736 0.648 0.73  0.739 0.699 0.786 0.735 0.773 0.785]
Standard_Deviation=0.038

Prediction_Confusion_Matrix:
[62|35]    	                      